### ISSUE TITLE
ggplot2 wrapper improvements

### ISSUE THREAD

**Issue by krassowski** [rpy2/rpy2#767]
The ggplot wrapper can be super useful, but it is not perfect yet. Some functions are missing and some functions would benefit better Python integration. I will try to describe some improvements that I would consider worthwhile below.

I think that the goal should be to demonstrate how to create nice wrappers and create a reasonably fidel wrappers (this is most of the examples from the ggplot documentation work with copy-paste + change of dots to underscores + addition of quotation marks where needed). Obviously the goal should not be to make every single possible usage of ggplot2 to work, but making sure the examples from official ggplot2 docs work after necessary adaptations (which could be automated in tests btw) would be a nice criterion for scope.

### 1) GGPlot constructor does not accept `mapping` argument

The most common usage of ggplot, appears to be `ggplot(data, mapping)`, but it is not currently supported.

Proposed solution draft:

```python
class GGPlot(robjects.vectors.ListVector):

    # ....

    @classmethod
    def new(cls, data, mapping=None):
        """ Constructor for the class GGplot. """
        data = conversion.py2rpy(data)
        res = cls(cls._constructor(data))
        if mapping:
            res = res + mapping
        return res
```

This should allow for:

```python
ggplot(mtcars, aes(x='disp', y='drat'))
```

### 2) `aes` does not accept positional arguments

```python
ggplot(mtcars, aes('disp', 'drat'))
```

is not currently possible.

### 3) [`labeller`](https://ggplot2.tidyverse.org/reference/labeller.html) and [`as_labeller`](https://ggplot2.tidyverse.org/reference/as_labeller.html) functions are missing.

`as_labeller()` is a simple function that maps data values to labels; as a first argument it accepts named string vector:

```R
to_string <- as_labeller(c(`0` = "Zero", `1` = "One"))
p + facet_wrap(~am, labeller = to_string)
```

or a function:

```R
appender <- function(string, suffix = "-foo") paste0(string, suffix)
p + facet_wrap(~am, labeller = as_labeller(appender))
```

Proposed wrapper implementation:

```python
from typing import Callable, Dict, Union

def as_labeller(x: Union[Dict[str, str], Callable[[str], str]], *args, **kwargs):
    if not callable(x):
        r_x = StrVector(list(x.values()))
        r_x.names = list(x.keys())
    else:
        r_x = x   # TODO (how to safely convert arguments and pass a function?)
    return ggplot2.as_labeller(r_x, *args, **kwargs)
```

Usage test 1:

```python
from rpy2.robjects.lib.ggplot2 import *
from rpy2.robjects.packages import importr, data
datasets = importr('datasets')
mtcars_env = data(datasets).fetch('mtcars')
mtcars = mtcars_env['mtcars']

# note: documentation uses ggplot(mtcars, aes('disp', 'drat')) but this is not possible because of point 1 and 2
p = ggplot(mtcars) + aes(x='disp', y='drat') + geom_point()
to_string = as_labeller({'0': 'Zero', '1': 'One'})
p + facet_wrap('~am', labeller=to_string)
```

Usage test 2:

```python
def appender(string, suffix="-foo"):
    return string + suffix  # TODO: may need to be vectorized

p + facet_wrap('~am', labeller=as_labeller(appender))
```

### 4) allow to pass dictionary for values and labels of `scale_*_manual`

Similarly to `as_labeller`, scales take two arguments that may be mappings (names correspond to keys, values to values) and thus python dictionary could be accepted as input format for those arguments. Currently `scale_fill_manual(values={'category one': 'red', 'category two': 'blue'})` raises:

`NotImplementedError: Conversion 'py2rpy' not defined for objects of type '<class 'dict'>'`

But ideally it should convert to named vector (as proposed in `as_labeller`); relevant documentation:

> values: a set of aesthetic values to map data values to. The values will be matched in order (usually alphabetical) with the limits of the scale, or with breaks if provided. If this is a named vector, then the values will be matched based on the names instead.

### 5) Other functions are missing

These functions can be defined as simple aliases and do not need any special wrappers:

- [x] xlab
- [x] ylab
- [x] guide_axis
- [x] guide_bins
- [x] guide_colorsteps()
- [x] guide_coluorsteps()
- [x] guide_none()

and many many more.

Two ideas to move forward:
- [ ] have a CI script that is triggered once in a while and imports ggplot and compares the namespace against the wrapper could be very useful. Of course one would not aim for 100% coverage because there are some things that are not relevant, but it would be great to catch changes between releases and other accidental omissions.
- [ ] replace the manual assignments with a generated code, this is replace this:

https://github.com/rpy2/rpy2/blob/57f8b65f0b0114619ef7be25536ecd19b043d5b2/rpy2/robjects/lib/ggplot2.py#L1116-L1118

with this:

```python
function_families_to_alias = {'guide'}
for name, func in vars(ggplot2).items():
    if name.split('_')[0] in function_families_to_alias:
        globals()[name] = func
```

**Comment by lgautier** [comment]
The R package `ggplot2` is a moving target, and the wrapper in `rpy2` can only chase the changes. Over the years I have experimented with dynamic generation; I can assure you that R and its packages will always make it break at some point. The question is whether going dynamic will create more issues than it solves.

Having a CI that compares namespace of the wrapper and namespace of the R package is an interesting idea. I am not sure about how much work is required to be a practical enough solution.

**Comment by lgautier** [comment]
I opened PR #768 about this.

**Comment by lgautier** [comment]
- 1/  and 2/ are in PR #768.
- You are already proposing an implementation for 3). Do you want to add it with unit tests?
- For 4/, named vector are ordered. Is this always the case with Python `dict` (I remember reading a discussion on python-dev about making vanilla `dict` ordered, I don't know what happened or whether this an implementation detail for CPython)
- For 5/, may the imported package is a good way to access all those functions for which there is no specific mapping needed?

**Comment by krassowski** [comment]
> 1/ and 2/ are in PR #768.

Thank you!

> You are already proposing an implementation for 3). Do you want to add it with unit tests?

I could not figure out how to make the function callback to work properly, it still needs work. I may have a look at it later (in a weeks time) - feel free to pick it up if you know a way and wish to contain it in a release sooner.

> For 4/, named vector are ordered. Is this always the case with Python dict (I remember reading a discussion on python-dev about making vanilla dict ordered, I don't know what happened or whether this an implementation detail for CPython)

Since Python 3.7 `dict` is insertion ordered; it was implementation detail of CPython in 3.6. I do not think that this is important though because the order of the string vector does not determine anything in the ggplot implementation (only the names and values matter) for values of discrete scales. It may matter in other cases though.

> For 5/, may the imported package is a good way to access all those functions for which there is no specific mapping needed?

Yes, this is obviously a fine workaround but would be nicer to have all under one namespace.

**Comment by lgautier** [comment]
> > 1/ and 2/ are in PR #768.
> 
> Thank you!
> 
> > You are already proposing an implementation for 3). Do you want to add it with unit tests?
> 
> I could not figure out how to make the function callback to work properly, it still needs work. I may have a look at it later (in a weeks time) - feel free to pick it up if you know a way and wish to contain it in a release sooner.

The callback must be an R function. There is a way to turn Python functions into R functions (`rinterface.rternalize()`). It might be better to first document how to do it than make the default though.
 
> > For 4/, named vector are ordered. Is this always the case with Python dict (I remember reading a discussion on python-dev about making vanilla dict ordered, I don't know what happened or whether this an implementation detail for CPython)
> 
> Since Python 3.7 `dict` is insertion ordered; it was implementation detail of CPython in 3.6. I do not think that this is important though because the order of the string vector does not determine anything in the ggplot implementation (only the names and values matter) for values of discrete scales. It may matter in other cases though.

Accepting any object with a method `items()` that returns (key, value) pairs might be a way to let the user decide what to use as input. 

> > For 5/, may the imported package is a good way to access all those functions for which there is no specific mapping needed?
> 
> Yes, this is obviously a fine workaround but would be nicer to have all under one namespace.

It would, but in a context of the R package's API evolving this allows a relatively quick and simple workaround (or may be the documented design choice that functions for which a Python wrapper would not add anything can be found there).

**Comment by lgautier** [comment]
Point 3/ is now added, as well as some of point 4/.

**Comment by lgautier** [comment]
While looking at implementing point 4/ I realized that it would be better to have rpy2 provide an helper function to cast a Python `Dict[str, str]` into an R named vector of strings.

**Comment by lgautier** [comment]
> While looking at implementing point 4/ I realized that it would be better to have rpy2 provide an helper function to cast a Python `Dict[str, str]` into an R named vector of strings.
> 
> This would work with the R functions `as_labeller` and `scale_*_manual`. Something like
> 
> ```python
> as_labeller(d2v({'a': 'b', 'c': 'd'}))
> ```

In the end the function that turn a Python `Dict[str, str]` into an R named vector of strings is called `dict2rvec()`
https://github.com/rpy2/rpy2/blob/master/rpy2/robjects/lib/ggplot2.py#L1342

All changes to improve the support of ggplot2 will be included in rpy2-3.4.3.
